### PR TITLE
Add support for expiring acknowledgements

### LINKED ISSUE #302: Acknowledge problem temporarily
I was using before OMD 1.10 + Thruk + Shinken.
Now I'm using OMD 2.90 labs + Thruk + Naemon.

With Shiken, I was able to acknowledge a problem during a limited period of time.
With Naemon, I don't have this choice.

### PR DESCRIPTION
**Pull request by roelvanmeer**
This adds support for acknowledgements that can expire.

I've split everything up in small commits. Please let me know if you'd like to have this differently, or if any improvements can be made. For one thing, I was unsure about the duplicate declaration of the handle_(host|service)_acknowledgement_expire_event functions. It might be better to put these in a header file.

I'm currently testing this out on our systems, but I'd like some preliminary feedback if possible.

Closes #302

### CODE REVIEW

**Comment by sni** [comment]
very nice. I'll try this out here as well.

**Review by jacobbaungard — COMMENTED** [review]
Thanks a lot for the contribution, and apologies I haven't been able to review this sooner. This is a nice feature. I did have reviewed and I think it looks really good. I added one minor comment.

Also it would be nice if it is possible to add a test where the expiration of the acknowledgement happens. Create an acknowledgement which expires shortly, check the acknowledgement is there, sleep a a little while, check the acknowledgement has been correctly expired.

@sni did you manage to do some testing of this?

File: src/naemon/xrddefault.c [hunk 1]
```diff
@@ -14,8 +14,13 @@
 #include "defaults.h"
 #include "nm_alloc.h"
 #include "broker.h"
+#include "events.h"
 #include <string.h>
 
+/* from commands.c */
+void handle_host_acknowledgement_expire_event(struct nm_event_execution_properties *evprop); /* removes an expired host acknowledgement */
```
[the remark this answers is not in the document]
Comment (roelvanmeer) [reply src/naemon/xrddefault.c:21]: I wasn't sure about the best way to solve this. Including commands.h includes a lot more definitions, which may be not preferable? I also considered creating a new header file, similar to downtime.h, called acknowledgement.h, and putting the definitions in there, but that means creating another file.
If you state your preference, I'll update the code accordingly.

**Comment by sni** [comment]
> @sni did you manage to do some testing of this?

not yet, i am still on vacation. Will do that next week.

File: src/naemon/xrddefault.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jacobbaungard) [reply src/naemon/xrddefault.c:21]: I definitely prefer including a large header file compared to having the function declarations twice. I think it's also OK to move `handle_host/service_acknowledgement_expire_event` etc to acknowledgement.c/h but then there's probably more functions we should move there, so then it becomes a bit more effort.

I would think say including `commands.h` is probably our best bet for now.

**Comment by jacobbaungard** [comment]
> Another way may be to test this the light way: schedule an ack that expires in two seconds, wait three, run the expire event manually, and check the results?

That sounds like a good way forward.

**Review by roelvanmeer — COMMENTED** [review]
This is just a stripped down copy of the test_commands.c file. The test works, but the way the event queue is emptied is not very robust now. If this looks promising, I would like some advice on how to progress.
Ideas: create a test function that empties the event queue, and run that before the test, so after we run the external command we should have to poll the event queue only once. If we do it that way, these tests can also be done in test_commands again, now that I think about it.

**Comment by roelvanmeer** [comment]
Update: function declaration has been de-duplicated, and the test has been added. I'd love to hear what you think.

File: t-tap/test_commands.c [hunk 2]
```diff
@@ -509,6 +510,30 @@ void test_host_commands(void)
...
+
+	target_host->current_state = STATE_DOWN;
+	ok(CMD_ERROR_OK == process_external_command1("[1234567890] ACKNOWLEDGE_HOST_PROBLEM_EXPIRE;host1;2;0;0;946681200;myself;expire in the past"), "core command: ACKNOWLEDGE_HOST_PROBLEM_EXPIRE");
+	ok(!target_host->problem_has_been_acknowledged, "ACKNOWLEDGE_HOST_PROBLEM_EXPIRE with past end_time doesn't acknowledge a host problem");
+	target_host->current_state = STATE_UP;
+
+	clear_event_queue();
+	target_host->current_state = STATE_DOWN;
+	time(&current_time);
+	nm_asprintf(&cmdstr, "[1234567890] ACKNOWLEDGE_HOST_PROBLEM_EXPIRE;host1;2;0;0;%llu;myself;expire in two seconds", (unsigned long long int)current_time+2);
```
Comment (jacobbaungard) [inline t-tap/test_commands.c:529]: Perhaps `1234567890` -> `current_time` now that we anyway have it.

**Review by jacobbaungard — COMMENTED** [review]
Great, thanks for adding the additional test. I think the code looks good, except for two small indentation issues, and another very minor comment.

If everything works fine following some manual testing, I think this should be fine to merge. I might be able to find some time late in the week to do so, otherwise perhaps sni can do it next week.

Thanks again for the contribution!

File: src/naemon/nebstructs.h [hunk 3]
```diff
@@ -470,6 +470,7 @@ typedef struct nebstruct_acknowledgement_struct {
 	int             persistent_comment;
 	int             notify_contacts;
 	void            *object_ptr;
+	time_t		end_time;
```
[the remark this answers is not in the document]
Comment (roelvanmeer) [reply src/naemon/nebstructs.h]: Fixed

File: src/naemon/objects_host.h [hunk 4]
```diff
@@ -77,6 +77,7 @@ struct host {
 	customvariablesmember *custom_variables;
 	int     problem_has_been_acknowledged;
 	int     acknowledgement_type;
+	time_t	acknowledgement_end_time;
```
[the remark this answers is not in the document]
Comment (roelvanmeer) [reply src/naemon/objects_host.h]: Fixed

File: t-tap/test_commands.c (hunk 2 above)
Comment (roelvanmeer) [reply t-tap/test_commands.c]: This isn't done in any of the other tests, but I did update the commit and removed the current_time variable; it's not used in other tests either.

**Review by sni — APPROVED** [review]
looks good, thanks a lot

**Comment by sni** [comment]
i had to rebase and push it manually. But it's merged now. Thanks a lot for this nice feature.

**Pull request closed without merging**